PHP 8.0 | Hooks::dispatch(): prevent potential fatal error

jrfnl opened the pull request:
### PHP 8.0 | Hooks::dispatch(): prevent potential fatal error

This commit adds a safeguard on the off-chance that the `Hooks::dispatch()` method would ever be called with an associative array for the parameters.

This prevents a potential fatal `Unknown named parameter ...` error in PHP 8.x.

Includes unit test.

Related to #533

### 🆕  Hooks::dispatch(): modernize dynamic function call

... by replacing the `call_user_func_array()` with a variable function call in combination with the spread operator.

Includes adding a dedicated test to safeguard that this change does not break the handling of references when passed as part of the `$parameters` to dispatch.

jrfnl committed:
PHP 8.0 | Hooks::dispatch(): prevent potential fatal error

This commit adds a safeguard on the off-chance that the `Hooks::dispatch()` method would ever be called with an associative array for the parameters.

This prevents a potential fatal `Unknown named parameter ...` error in PHP 8.x.

Includes unit test.

schlessera commented on src/Hooks.php:
```diff
@@ -83,6 +83,11 @@ public function dispatch($hook, $parameters = array()) {
+
 		foreach ($this->hooks[$hook] as $priority => $hooked) {
 			foreach ($hooked as $callback) {
 				call_user_func_array($callback, $parameters);
```
Why are we not using variadic arguments here instead?
```suggestion
				$callback(...$parameters);
```

jrfnl replied:
> Why are we not using variadic arguments here instead?

Historically because it wasn't possible previously as the minimum PHP version did not allow for the spread operator ;-)

As for changing it now... I'm a bit wary of doing so as the behaviour will change with regards to references and the method is called quite often with variables being passed by reference.

Also see https://core.trac.wordpress.org/ticket/47678#comment:31 - section 3.

jrfnl replied:
Then again... `array_values()` may already break the references...

schlessera replied:
Yes, that's what I thought too. If we are using `array_values()`, it will be pretty much the same thing as using the variadic arguments operator.

schlessera replied:
Well, I'm not 100% sure, I think this needs to be tested.

jrfnl replied:
I've tested this and added a unit test to safeguard this as well.

The test passed prior to the code changes and still passes with both the code changes in place, so I think we're good.

jrfnl replied:
(to my surprise, I may add)

jrfnl committed:
Hooks::dispatch(): modernize dynamic function call

... by replacing the `call_user_func_array()` with a variable function call in combination with the spread operator.

Includes adding a dedicated test to safeguard that this change does not break the handling of references when passed as part of the `$parameters` to dispatch.

jrfnl commented:
FYI: re-pushed commit has no functional changes, just added the `$message` parameter to the assertions as there are multiple assertions in the `testDispatchHandlesParametersPassedByReference()` test.
